rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/dap/RascalDebugAdapter.java

```diff
@@ -0,0 +1,420 @@
+                    scopeModule.setPresentationHint("module");
+                    scopeModule.setExpensive(module.getFrameVariables().size()>expensiveScopeMinSize);
+                    scopeModule.setVariablesReference(suspendedState.addScope(module));
+                    scopes.add(scopeModule);
+                }
+            }
+
+            response.setScopes(scopes.toArray(new Scope[scopes.size()]));
+            return response;
+        });
+    }
+
+    @Override
+    public CompletableFuture<VariablesResponse> variables(VariablesArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            int reference = args.getVariablesReference();
+            int minIndex = args.getStart() != null ? args.getStart() : 0;
+            int maxCount = args.getCount() != null ? args.getCount() : -1;
+
+            VariablesResponse response = new VariablesResponse();
+            List<RascalVariable> variables = suspendedState.getVariables(reference, minIndex, maxCount);
+            Variable[] variablesResponse = new Variable[variables.size()];
+            int i = 0;
+            for(RascalVariable var : variables){
+                Variable variable = new Variable();
+                variable.setName(var.getName());
+                variable.setType(var.getType().toString());
+                variable.setValue(var.getDisplayValue());
+                variable.setVariablesReference(var.getReferenceID());
+                variable.setNamedVariables(var.getNamedVariables());
+                variable.setIndexedVariables(var.getIndexedVariables());
+                variablesResponse[i] = variable;
+                i++;
+            }
+
+            response.setVariables(variablesResponse);
+            return response;
+        });
+    }
+
+    @Override
+    public CompletableFuture<ContinueResponse> continue_(ContinueArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            ContinueResponse response = new ContinueResponse();
+            response.setAllThreadsContinued(true);
+
+            debugHandler.processMessage(DebugMessageFactory.requestResumption());
+
+            return response;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> next(NextArguments args) {
+        return CompletableFuture.supplyAsync(() -> {
+            debugHandler.processMessage(DebugMessageFactory.requestStepOver());
+            return null;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> disconnect(DisconnectArguments args) {
+        new java.lang.Thread(new Runnable() {
+            public void run() {
+                debugHandler.processMessage(DebugMessageFactory.requestTermination());
+                if(suspendedState.isSuspended()){
+                    debugHandler.processMessage(DebugMessageFactory.requestResumption());
+                }
+            }
+        }).start();
```

Comment:
maybe add some comments about why this is not done inside of a completed future? I'm guessing this is about shutdown logic? to prevent deadlocky stuff?